dna/modify.go

```diff
@@ -124,6 +130,17 @@ func Delete(seq []Base, delStart int64, delEnd int64) []Base {
 	return append(seq[:delStart], seq[delEnd:]...)
 }
 
+// DeleteStable is identical to Delete, but conserves the
+// underlying memory of the input destSeq slice
+func DeleteStable(seq *[]Base, delStart int, delEnd int) error {
+	if delStart >= delEnd || delStart < 0 || delEnd > len(*seq) {
+		return ErrInvalidPosition
+	}
+	*seq = (*seq)[:len(*seq)-(delEnd-delStart)]                // decrease the LEN of the slice to by num of deleted bases
+	copy((*seq)[delStart:cap(*seq)], (*seq)[delEnd:cap(*seq)]) // shift bases in slice to remove deleted bases
```

Comment:
Consider the following:
`Cdna := make([]Base, 4, 5)`
`Cdna = dna.StringToBases("ACTG") // cDNA sequence is: ACTG`
`CodingSeq = Cdna[1:3] // CDS is: CT`

The goal of this function is the following
`InsertStable(CodingSeq, 1, []Base{A})`
Expected cDNA sequence is: ACATG
Expected CDS sequence is: CAT

If I only copy to the len and not the cap then the result would be
Actual cDNA sequence is: ACATA <- last base is nil value
Actual CDS sequence is: CAT

The start/end of the `Cdna` slice will need to be adjusted after the insertion, but the fidelity of the entire sequence array is maintained.

Reply:
I think I understand the use case now.  Maybe this is specific to an insertion in your gene struct and does not smoothly generalize to any dna seq?
Does it work smoother if the insertion would be executed on the cDNA instead of the CDS?

Reply:
It does work smoothly if done on the cDNA and not the CDS. However, I think the use case for the IndelStable functions will be manipulating a slice where different variables hold a re-slice of the same underlying array. With multiple re-slices of an array the only way to safely make an indel would be to perform the function on an a slice that is a superset of all other slices (cDNA in this case). To maintain the ability to manipulate a subset slice I think we need to copy to the cap of the slice. I am not sure whether this generalizes to all []dna.Base. I think it depends on the use case. I would be happy to move these to private functions within the gene package if people think these are too specific.

Reply:
Maybe this is worth a quick discussion at our 4pm meeting?
I am thinking that preserving data between len and cap is pretty specific to this use case and could be computationally expensive.  Maybe there is a way to call the insert function on cDNA?  It seems like even if the underlying array of cDNA was changed by calling insert on CDS, we would still need to modify the cDNA slice to understand that its length needed to be updated, since the last base of the cDNA is now either closer or farther from the start base.  Since we need access to that struct anyway, maybe there is a way to call this function on that one?
I don't want to hold up your progress though.  I think this is a relatively minor point.  Maybe just make a detailed note in InsertStable documentation that it copies all seq between len and cap as well?

Reply:
Thank you Craig! That makes much more sense. I was thinking about it wrong, totally forgot we already made the total number of slices longer in the code on the line directly above.